File: 2-hybrid-search-with-rrf-reranking/0-typescript/src/hybrid/rrf.ts
import { FusedHit, UpstreamHit } from './types';

/**
 * Reciprocal Rank Fusion (Cormack et al., 2009).
 *
 * For every list, a document at 1-based rank `r` contributes `1 / (k + r)`. The
 * contributions are SUMMED across lists per document, so a document ranked decently
 * in BOTH lists collects two terms and rises above a document ranked first in only
 * one list. The formula reads only `rank` — the raw BM25 and cosine scores never
 * meet, so no score normalization is needed.
 */
export function fuseByRrf(
  bm25: UpstreamHit[],
  vector: UpstreamHit[],
  k: number,
  limit: number,
): FusedHit[] {
  const acc = new Map<string, FusedHit>();

  const accumulate = (list: UpstreamHit[], isBm25: boolean): void => {
    list.forEach((hit, idx) => {
      const rank = idx + 1; // 1-based: the first element has rank 1.
      const contribution = 1 / (k + rank);
      const existing = acc.get(hit.id);
      if (existing) {
        // A doc present in BOTH lists accumulates two contributions -> consensus boost.
        existing.rrfScore += contribution;
        if (isBm25) existing.sources.bm25Rank = rank;
        else existing.sources.vectorRank = rank;
      } else {
        acc.set(hit.id, {
          id: hit.id,
          content: hit.content,
          rrfScore: contribution,
          sources: {
            bm25Rank: isBm25 ? rank : null,
            vectorRank: isBm25 ? null : rank,
          },
        });
      }
    });
  };

  accumulate(bm25, true);
  accumulate(vector, false);

  // Sort by fused score descending; consensus documents naturally rise to the top.
  return [...acc.values()]
    .sort((a, b) => b.rrfScore - a.rrfScore)
    .slice(0, limit);
}

File: 2-hybrid-search-with-rrf-reranking/1-java/src/main/java/com/starci/hybrid/Rrf.java
package com.starci.hybrid;

import com.starci.hybrid.Models.Doc;
import com.starci.hybrid.Models.Hit;
import com.starci.hybrid.Models.Sources;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** Reciprocal Rank Fusion (Cormack et al., 2009). */
public final class Rrf {
    private Rrf() {}

    private static final class Acc {
        final String id;
        final String content;
        double rrfScore;
        Integer bm25Rank;
        Integer vectorRank;

        Acc(String id, String content) {
            this.id = id;
            this.content = content;
        }
    }

    /**
     * Fuse two ranked lists by rank position. A document at 1-based rank r contributes
     * 1/(k+r); contributions are summed across lists, so a document present in BOTH lists
     * gains a consensus boost. Only rank is read — no score normalization is needed.
     */
    public static List<Hit> fuse(List<Doc> bm25, List<Doc> vector, int k, int limit) {
        Map<String, Acc> acc = new LinkedHashMap<>();

        accumulate(acc, bm25, k, true);
        accumulate(acc, vector, k, false);

        List<Acc> values = new ArrayList<>(acc.values());
        values.sort(Comparator.comparingDouble((Acc a) -> a.rrfScore).reversed());

        List<Hit> out = new ArrayList<>();
        for (int i = 0; i < values.size() && i < limit; i++) {
            Acc a = values.get(i);
            out.add(new Hit(a.id, a.content, a.rrfScore, new Sources(a.bm25Rank, a.vectorRank)));
        }
        return out;
    }

    private static void accumulate(Map<String, Acc> acc, List<Doc> docs, int k, boolean isBm25) {
        for (int i = 0; i < docs.size(); i++) {
            Doc d = docs.get(i);
            int rank = i + 1; // 1-based
            Acc a = acc.computeIfAbsent(d.id(), id -> new Acc(d.id(), d.content()));
            a.rrfScore += 1.0 / (k + rank);
            if (isBm25) a.bm25Rank = rank;
            else a.vectorRank = rank;
        }
    }
}
